zeppelin-test/src/main/java/org/apache/zeppelin/test/DownloadUtils.java

```diff
@@ -0,0 +1,537 @@
+        }
+
+        // write file content
+        try (FileOutputStream fos = new FileOutputStream(newFile)) {
+          int len;
+          while ((len = zis.read(buffer)) > 0) {
+            fos.write(buffer, 0, len);
+          }
+        }
+        // Change permissions and metadata
+        if (newFile.getParentFile().getName().contains("bin")
+            && !newFile.setExecutable(true, false)) {
+          LOGGER.info("Setting file {} to executable failed", newFile);
+        }
+        if (!newFile.setLastModified(zipEntry.getLastModifiedTime().toMillis())) {
+          LOGGER.info("Setting last modified date to file {} failed", newFile);
+        }
+      }
+      zipEntry = zis.getNextEntry();
+    }
+    zis.closeEntry();
+  }
+
+  public static String downloadFlink(String flinkVersion, String scalaVersion) {
+    File flinkDownloadFolder = new File(downloadFolder, "flink");
+    flinkDownloadFolder.mkdir();
+    File targetFlinkHomeFolder = new File(flinkDownloadFolder, "flink-" + flinkVersion);
+    if (targetFlinkHomeFolder.exists()) {
+      LOGGER.info("Skip to download Flink {}_{} as it is already downloaded.", flinkVersion,
+          scalaVersion);
+      return targetFlinkHomeFolder.getAbsolutePath();
+    }
+    File flinkTGZ = new File(flinkDownloadFolder,
+        "flink-" + flinkVersion + "-bin-scala_" + scalaVersion + ".tgz");
+    try {
+      URL mirrorURL = new URL(
+          IOUtils.toString(new URL(MIRROR_URL), StandardCharsets.UTF_8) + generateDownloadURL(
+              "flink", flinkVersion, "-bin-scala_" + scalaVersion + ".tgz", "flink"));
+      URL archiveURL = new URL(ARCHIVE_URL + generateDownloadURL(
+          "flink", flinkVersion, "-bin-scala_" + scalaVersion + ".tgz", "flink"));
+      LOGGER.info("Download Flink {}_{}", flinkVersion, scalaVersion);
+      download(new DownloadRequest(mirrorURL, archiveURL), flinkTGZ);
+      ProgressBarBuilder pbb = new ProgressBarBuilder()
+          .setTaskName("Unarchiv Flink")
+          .setUnit("MiB", 1048576) // setting the progress bar to use MiB as the unit
+          .setStyle(ProgressBarStyle.ASCII)
+          .setUpdateIntervalMillis(1000)
+          .setConsumer(new DelegatingProgressBarConsumer(LOGGER::info));
+      try (
+          InputStream fis = Files.newInputStream(flinkTGZ.toPath());
+          InputStream pbis = ProgressBar.wrap(fis, pbb);
+          InputStream bis = new BufferedInputStream(pbis);
+          InputStream gzis = new GzipCompressorInputStream(bis);
+          ArchiveInputStream<TarArchiveEntry> o = new TarArchiveInputStream(gzis)) {
+        LOGGER.info("Unarchive Flink {}_{} to {}", flinkVersion, scalaVersion,
+            targetFlinkHomeFolder);
+        unarchive(o, targetFlinkHomeFolder, 1);
+        LOGGER.info("Unarchive Flink done");
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to download flink", e);
+    }
+
+
+    // download other dependencies for running flink with yarn and hive
```

Comment:
I touched this part in https://github.com/apache/zeppelin/commit/fa6e3ee3b83aa5ffc070557780ec111386b8c951, could you please sync the change?

Reply:
Many thanks for the hint. I have adjusted the download paths.